File: chapter 07/recipe 7.1/DniNetInOutVB/DniNetInOut.vb
Imports System.Runtime.InteropServices

Public Interface IDirectionTester
    Sub IntInOnly(ByVal paramA As Integer, _
        ByVal paramB As Integer)
    Sub IntInAndOut(ByVal paramA As Integer, _
        <[In](), Out()> ByVal paramB As Integer)
    Sub IntByRef(ByVal paramA As Integer, _
        ByRef paramB As Integer)
    Sub IntByRefInOnly(ByVal paramA As Integer, _
        <[In]()> ByRef paramB As Integer)
    'sub IntOut(int paramA, out int paramB);
    Sub StringsInOnly(ByVal paramA As String, _
        ByVal paramB As String)
    Sub StringsInAndOut(ByVal paramA As String, _
        <[In](), Out()> ByVal paramB As String)
    Sub StringsByRef(ByVal paramA As String, _
        ByRef paramB As String)
    Sub StringsByRefInOnly(ByVal paramA As String, _
        <[In]()> ByRef paramB As String)
    'sub StringsOut(string paramA, out string paramB);
    Sub DateTimeInOnly(ByVal paramA As DateTime, _
        ByVal paramB As DateTime)
    Sub DateTimeInAndOut(ByVal paramA As DateTime, _
        <[In](), Out()> ByVal paramB As DateTime)
    Sub DateTimeByRef(ByVal paramA As DateTime, _
        ByRef paramB As DateTime)
    Sub DateTimeByRefInOnly(ByVal paramA As DateTime, _
        <[In]()> ByRef paramB As DateTime)
    'sub DateTimeOut(DateTime paramA, out DateTime paramB);
End Interface

<ClassInterface(ClassInterfaceType.None)> _
Public Class DniNetInOutObj
    Implements IDirectionTester

    Public Sub DateTimeByRef(ByVal paramA As Date, _
        ByRef paramB As Date) _
        Implements IDirectionTester.DateTimeByRef
        paramB = paramA
    End Sub

    Public Sub DateTimeByRefInOnly(ByVal paramA As Date, _
        ByRef paramB As Date) _
        Implements IDirectionTester.DateTimeByRefInOnly
        paramB = paramA
    End Sub

    Public Sub DateTimeInAndOut(ByVal paramA As Date, _
        ByVal paramB As Date) _
        Implements IDirectionTester.DateTimeInAndOut
        paramB = paramA
    End Sub

    Public Sub DateTimeInOnly(ByVal paramA As Date, _
        ByVal paramB As Date) _
        Implements IDirectionTester.DateTimeInOnly
        paramB = paramA
    End Sub

    Public Sub IntByRef(ByVal paramA As Integer, _
        ByRef paramB As Integer) _
        Implements IDirectionTester.IntByRef
        paramB = paramA
    End Sub

    Public Sub IntByRefInOnly(ByVal paramA As Integer, _
        ByRef paramB As Integer) _
        Implements IDirectionTester.IntByRefInOnly
        paramB = paramA
    End Sub

    Public Sub IntInAndOut(ByVal paramA As Integer, _
        ByVal paramB As Integer) _
        Implements IDirectionTester.IntInAndOut
        paramB = paramA
    End Sub

    Public Sub IntInOnly(ByVal paramA As Integer, _
        ByVal paramB As Integer) _
        Implements IDirectionTester.IntInOnly
        paramB = paramA
    End Sub

    Public Sub StringsByRef(ByVal paramA As String, _
        ByRef paramB As String) _
        Implements IDirectionTester.StringsByRef
        paramB = paramA
    End Sub

    Public Sub StringsByRefInOnly(ByVal paramA As String, _
        ByRef paramB As String) _
        Implements IDirectionTester.StringsByRefInOnly
        paramB = paramA
    End Sub

    Public Sub StringsInAndOut(ByVal paramA As String, _
        ByVal paramB As String) _
        Implements IDirectionTester.StringsInAndOut
        paramB = paramA
    End Sub

    Public Sub StringsInOnly(ByVal paramA As String, _
        ByVal paramB As String) _
        Implements IDirectionTester.StringsInOnly
        paramB = paramA
    End Sub
End Class

File: chapter 07/recipe 7.1/DniNetInOut/DniNetInOut.cs
using System;
using System.Runtime.InteropServices;

namespace DniNetInOut
{
	public interface IDirectionTester
	{
		void IntInOnly(int paramA, int paramB);
		void IntInAndOut(int paramA, [In,Out] int paramB);
		void IntByRef(int paramA, ref int paramB);
		void IntByRefInOnly(int paramA, [In] ref int paramB);
		void IntOut(int paramA, out int paramB);
		void StringsInOnly(string paramA, string paramB);
		void StringsInAndOut(string paramA, [In,Out] string paramB);
		void StringsByRef(string paramA, ref string paramB);
		void StringsByRefInOnly(string paramA, [In] ref string paramB);
		void StringsOut(string paramA, out string paramB);
		void DateTimeInOnly(DateTime paramA, DateTime paramB);
		void DateTimeInAndOut(DateTime paramA, [In,Out] DateTime paramB);
		void DateTimeByRef(DateTime paramA, ref DateTime paramB);
		void DateTimeByRefInOnly(DateTime paramA, [In] ref DateTime paramB);
		void DateTimeOut(DateTime paramA, out DateTime paramB);
	}

	[ClassInterface(ClassInterfaceType.None)]
	public class DniNetInOutObj : IDirectionTester 
	{
		//blittable type - by value
		public void IntInOnly(int paramA, int paramB)
		{
			paramB = paramA;
		}
		//blittable type - by value, with out attribute
		public void IntInAndOut(int paramA, int paramB)
		{
			paramB = paramA;
		}
		//blittable type - by ref
		public void IntByRef(int paramA, ref int paramB)
		{
			paramB = paramA;
		}
		//blittable type - by ref, with in attribute
		public void IntByRefInOnly(int paramA, ref int paramB)
		{
			paramB = paramA;
		}
		//blittable type - out
		public void IntOut(int paramA, out int paramB)
		{
			paramB = paramA;
		}

		//reference type - by value
		public void StringsInOnly(string paramA, string paramB)
		{
			paramB = paramA;
		}
		//reference type - by value, with out attribute
		public void StringsInAndOut(string paramA, string paramB)
		{
			paramB = paramA;
		}
		//reference type - by ref
		public void StringsByRef(string paramA, ref string paramB)
		{
			paramB = paramA;
		}
		//reference type - by ref, with in attribute
		public void StringsByRefInOnly(
			string paramA, ref string paramB)
		{
			paramB = paramA;
		}
		//reference type - out
		public void StringsOut(string paramA, out string paramB)
		{
			paramB = paramA;
		}

		//value type - by value
		public void DateTimeInOnly(
			DateTime paramA, DateTime paramB)
		{
			paramB = paramA;
		}
		//value type - by value, with out attribute
		public void DateTimeInAndOut(
			DateTime paramA, DateTime paramB)
		{
			paramB = paramA;
		}
		//value type - by ref
		public void DateTimeByRef(
			DateTime paramA, ref DateTime paramB)
		{
			paramB = paramA;
		}
		//value type - by ref, with in attribute
		public void DateTimeByRefInOnly(
			DateTime paramA, ref DateTime paramB)
		{
			paramB = paramA;
		}
		//value type - out
		public void DateTimeOut(
			DateTime paramA, out DateTime paramB)
		{
			paramB = paramA;
		}

	}
}
